### PR TITLE
Unix socket support for subgraphs

### LINKED ISSUE #3504: Send traffic to subgraphs via unix domain socket
(Filing this request after asking in Slack and being directed here by @lennyburdette)

**Is your feature request related to a problem? Please describe.**
We have subgraphs we send traffic to by communicating with a sidecar via a unix domain socket. In the gateway this worked for us because we wrote a datasource using a networking package that supported this.

**Describe the solution you'd like**
The ability to configure egress for certain subgraphs to be sent over a particular UDS.

**Describe alternatives you've considered**
N/A

**Additional context**
It looks like `hyper::client` doesn't directly support sending egress via UDS. In https://github.com/hyperium/hyper/issues/126 someone referenced a separate crate they developed that does support it https://github.com/softprops/hyperlocal

### PR DESCRIPTION
**Pull request by Geal**
Fix #3504

This implements support for Unix sockets in subgraph connections by specifying URLs in the `unix:///path/to/router.sock` format.

Implementation details:
- the Uri type used throughout the router does not actually support the `unix://` scheme, and there is no real standard for it, so the library we use here hides the socket path by hex encoding it in the authority part of the Uri
- this only supports stream sockets
- decompression is supported
- TLS is not supported

Note: I'd like to add it to the doc but can't decide on the right place to put it. Any suggestions?

TODO:
- [x] update the override_subgraph_url plugin to support it

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/src/spec/schema.rs [hunk 1]
```diff
@@ -116,8 +116,21 @@ impl Schema {
                 if url.is_empty() {
                     return Err(SchemaError::MissingSubgraphUrl(name.to_string()));
                 }
-                let url = Uri::from_str(url)
-                    .map_err(|err| SchemaError::UrlParse(name.to_string(), err))?;
+                // there is noi standard for unix socket URLs apparently
```
Comment (lennyburdette) [inline apollo-router/src/spec/schema.rs:119]:
```suggestion
                // there is no standard for unix socket URLs apparently
```

File: apollo-router/src/spec/schema.rs [hunk 2]
```diff
@@ -116,8 +116,21 @@ impl Schema {
...
+                let url = if url.starts_with("unix://") {
+                    println!("authority: {}", &url[7..]);
```
Comment (lennyburdette) [inline apollo-router/src/spec/schema.rs:121]:
```suggestion
```

File: apollo-router/src/spec/schema.rs [hunk 3]
```diff
@@ -116,8 +116,21 @@ impl Schema {
...
+                    // there is no specified format for unix socket URLs (cf https://github.com/whatwg/url/issues/577)
+                    // so a unix:// URL will not be parsed by http::Uri
+                    // To fix that, hyperlocal came up with its own Uri type that can be converted to http::Uri.
+                    // It hides the socket path in a hex encoded authority that the unix socket connector will
+                    // know how to decode
+                    let u = hyperlocal::Uri::new(&url[7..], "/").into();
+                    println!("generated Uri = {u:?} from '{url}'");
```
Comment (lennyburdette) [inline apollo-router/src/spec/schema.rs:128]:
```suggestion
```

**Review by lennyburdette — APPROVED** [review]
awesome!

File: apollo-router/src/spec/schema.rs (hunk 2 above)
Comment (Geal) [reply apollo-router/src/spec/schema.rs]: fixed in [512b54b](https://github.com/apollographql/router/pull/4757/commits/512b54b3a36c6f13168081965eb6a33052b0d482)

File: apollo-router/src/spec/schema.rs (hunk 3 above)
Comment (Geal) [reply apollo-router/src/spec/schema.rs]: fixed in [512b54b](https://github.com/apollographql/router/pull/4757/commits/512b54b3a36c6f13168081965eb6a33052b0d482)

**Comment by abernix** [comment]
@shorgi Do you have suggestions on where this can land in the docs?

**Comment by Geal** [comment]
it is a new subgraph URL, and it will be supported by override_subgraph_url. But also, it could come from the supergraph schema

**Comment by Geal** [comment]
ok, it's mentioned in the subgraph routing urls section now https://github.com/apollographql/router/pull/4757/files#diff-69d7e3fff212844102e67091fcc275b06d67899a25df2c16f9f5a78bfd7e6444R517

**Comment by Geal** [comment]
it needs a few more fixes to make sure builds are passing on Windows (where unix sockets don't exist) then I'll merge it

**Pull request merged**